hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -82,262 +80,268 @@ object InferPType {
-        val seqEnv = resEnv.bind(valueName -> a.pType2.asInstanceOf[PArray].elementType)
+
+        val resEnv = env.bind(acc.map { case (name, accIR) => (name, accIR._pType2)}: _*)
+        val seqEnv = resEnv.bind(valueName -> a._pType2.asInstanceOf[PArray].elementType)
         seq.foreach(InferPType(_, seqEnv))
         InferPType(res, resEnv)
-        res.pType2.setRequired(res.pType2.required && a.pType2.required)
+        res._pType2.setRequired(res._pType2.required && a._pType2.required)
       case ArrayScan(a, zero, accumName, valueName, body) => {
         InferPType(zero, env)
 
         InferPType(a, env)
-        InferPType(body, env.bind(accumName -> zero.pType2, valueName -> a.pType2.asInstanceOf[PArray].elementType))
-        assert(body.pType2 isOfType zero.pType2)
+        InferPType(body, env.bind(accumName -> zero._pType2, valueName -> a._pType2.asInstanceOf[PArray].elementType))
+        assert(body._pType2 isOfType zero._pType2)
 
-        val elementPType = zero.pType2.setRequired(body.pType2.required && zero.pType2.required)
-        coerce[PStreamable](a.pType2).copyStreamable(elementPType, a.pType2.required)
+        val elementPType = zero._pType2.setRequired(body._pType2.required && zero._pType2.required)
+        coerce[PStreamable](a._pType2).copyStreamable(elementPType, a._pType2.required)
       }
       case ArrayLeftJoinDistinct(lIR, rIR, lName, rName, compare, join) => {
         InferPType(lIR, env)
         InferPType(rIR, env)
 
-        InferPType(join, env.bind(lName -> lIR.pType2.asInstanceOf[PArray].elementType, rName -> rIR.pType2.asInstanceOf[PArray].elementType))
+        InferPType(join, env.bind(lName -> lIR._pType2.asInstanceOf[PArray].elementType, rName -> rIR._pType2.asInstanceOf[PArray].elementType))
 
-        PArray(join.pType2, lIR.pType2.required)
+        PArray(join._pType2, lIR._pType2.required)
       }
       case NDArrayShape(nd) => {
         InferPType(nd, env)
-        PTuple(nd.pType2.required, IndexedSeq.tabulate(nd.pType2.asInstanceOf[PNDArray].nDims)(_ => PInt64(true)):_*)
+        PTuple(nd._pType2.required, IndexedSeq.tabulate(nd._pType2.asInstanceOf[PNDArray].nDims)(_ => PInt64(true)):_*)
       }
       case NDArrayReshape(nd, shape) => {
         InferPType(nd, env)
         InferPType(shape, env)
 
-        PNDArray(coerce[PNDArray](nd.pType2).elementType, shape.pType2.asInstanceOf[TTuple].size, nd.pType2.required)
+        PNDArray(coerce[PNDArray](nd._pType2).elementType, shape._pType2.asInstanceOf[PTuple].size, nd._pType2.required)
       }
       case NDArrayMap(nd, name, body) => {
         InferPType(nd, env)
-        InferPType(body, env.bind(name, nd.pType2))
+        val ndPType = nd._pType2.asInstanceOf[PNDArray]
+        InferPType(body, env.bind(name -> ndPType.elementType))
 
-        PNDArray(body.pType2, coerce[PNDArray](nd.pType2).nDims, nd.pType2.required)
+        PNDArray(body._pType2, ndPType.nDims, body._pType2.required)
```

Comment:
TODO: figure out whether I should keep requiredeness of PNDArray here (and NDArrayMap2) as that of body, as done in ArrayMap, or on the left hand operand, as currently in master. If none of the elements of the resulting array are required, do we construct the array?